File: dropwizard/src/main/java/nl/jaapcoomans/microframeworks/dropwizard/TodoRestController.java
package nl.jaapcoomans.microframeworks.dropwizard;

import nl.jaapcoomans.demo.microframeworks.todo.api.CreateTodoCommand;
import nl.jaapcoomans.demo.microframeworks.todo.api.TodoDTO;
import nl.jaapcoomans.demo.microframeworks.todo.domain.PartialTodo;
import nl.jaapcoomans.demo.microframeworks.todo.domain.Todo;
import nl.jaapcoomans.demo.microframeworks.todo.domain.TodoService;

import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.PATCH;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.util.List;
import java.util.UUID;
import java.util.stream.Collectors;

@Path("/todos")
public class TodoRestController {
    private TodoService todoService;
    private String baseUrl;

    TodoRestController(TodoService todoService, String baseUrl) {
        this.todoService = todoService;
        this.baseUrl = baseUrl;
    }

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    public List<TodoDTO> getAll() {
        return this.todoService.findAll().stream()
                .map(this::wrap)
                .collect(Collectors.toList());
    }

    @GET
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    public Response getTodo(@PathParam("id") String id) {
        return this.todoService.findById(UUID.fromString(id))
                .map(this::wrap)
                .map(body -> Response.ok(body).build())
                .orElse(Response.status(Response.Status.NOT_FOUND).build());
    }

    @POST
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public TodoDTO createTodo(CreateTodoCommand command) {
        Todo result = this.todoService.createNewTodo(command.getTitle(), command.getOrder());
        return this.wrap(result);
    }

    @DELETE
    public Response deleteAll() {
        this.todoService.deleteAll();
        return Response.noContent().build();
    }

    @PATCH
    @Path("/{id}")
    @Produces(MediaType.APPLICATION_JSON)
    @Consumes(MediaType.APPLICATION_JSON)
    public TodoDTO patchTodo(@PathParam("id") String id, PartialTodo command) {
        return this.todoService.updateTodo(UUID.fromString(id), command)
                .map(this::wrap)
                .orElse(null);
    }

    @DELETE
    @Path("/{id}")
    public Response deleteTodo(@PathParam("id") UUID id) {
        this.todoService.delete(id);
        return Response.noContent().build();
    }

    private TodoDTO wrap(Todo todo) {
        return TodoDTO.from(todo, createUrl(todo));
    }

    private String createUrl(Todo todo) {
        return this.baseUrl + "/" + todo.getId().toString();
    }
}

File: http4k/src/main/kotlin/nl/jaapcoomans/demo/microframeworks/http4k/TodoRestController.kt
package nl.jaapcoomans.demo.microframeworks.http4k

import nl.jaapcoomans.demo.microframeworks.todo.api.CreateTodoCommand
import nl.jaapcoomans.demo.microframeworks.todo.api.TodoDTO
import nl.jaapcoomans.demo.microframeworks.todo.domain.PartialTodo
import nl.jaapcoomans.demo.microframeworks.todo.domain.Todo
import nl.jaapcoomans.demo.microframeworks.todo.domain.TodoService
import org.http4k.core.ContentType
import org.http4k.core.Request
import org.http4k.core.Response
import org.http4k.core.Status
import org.http4k.format.Jackson
import org.http4k.format.Jackson.asJsonObject
import org.http4k.routing.path
import java.util.UUID

class TodoRestController(val todoService: TodoService) {
    private val json = Jackson

    fun getAll(req: Request): Response {
        val todos = todoService.findAll().map { it.wrap() }
        return Response(Status.OK)
                .header("Content-Type", ContentType.APPLICATION_JSON.value)
                .body(todos.asJsonObject().toPrettyString())
    }

    fun createTodo(req: Request): Response {
        val command = json.asA(req.bodyString(), CreateTodoCommand::class)

        val todo = todoService.createNewTodo(command.title, command.order)

        return Response(Status.OK)
                .header("Content-Type", ContentType.APPLICATION_JSON.value)
                .body(todo.wrap().asJsonObject().toPrettyString())
    }

    fun getTodo(req: Request): Response {
        val id = req.path("id")
        val todo = todoService.findById(UUID.fromString(id))
        return if (todo.isPresent) {
            Response(Status.OK)
                    .header("Content-Type", ContentType.APPLICATION_JSON.value)
                    .body(todo.get().wrap().asJsonObject().toPrettyString())
        } else {
            Response(Status.NOT_FOUND)
        }
    }

    fun patchTodo(req: Request): Response {
        val id = req.path("id")
        val command = json.asA(req.bodyString(), PartialTodo::class)

        val todo = todoService.updateTodo(UUID.fromString(id), command)
        return if (todo.isPresent) {
            Response(Status.OK)
                    .header("Content-Type", ContentType.APPLICATION_JSON.value)
                    .body(todo.get().wrap().asJsonObject().toPrettyString())
        } else {
            Response(Status.NOT_FOUND)
        }
    }

    fun deleteTodo(req: Request): Response {
        val id = req.path("id")
        todoService.delete(UUID.fromString(id))
        return Response(Status.NO_CONTENT)
    }

    fun deleteAll(req: Request): Response {
        todoService.deleteAll()
        return Response(Status.NO_CONTENT)
    }

    private fun Todo.wrap(): TodoDTO {
        return TodoDTO.from(this, createUrl())
    }

    private fun Todo.createUrl(): String {
        return "http://localhost:8080/todos/$id"
    }
}